### PR TITLE
Integrate llvm-project to 9c6a2f29660b and bump dependencies.

### PR DESCRIPTION
**Pull request by dcaballe**
Integrate llvm-project and bump dependencies.
    
      * llvm-project: 9c6a2f29660b
      * mlir-hlo: 6cc16a2eae09
      * tensorflow: fe15774895bb

### CODE REVIEW

**Comment by dcaballe** [comment]
Could anyone please try to reproduce the CI errors? These ones in particular:
```
ERROR: /tmpfs/src/github/iree/llvm-external-projects/iree-dialects/BUILD:277:11: Compiling llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-6.0/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG ... (remaining 212 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1595:5: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
    parser.parseAttribute(valueAttr, TileOp::getTiledDimAttrName(),
    ^~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
I even installed clang-6.0 and tried a few things but I'm not getting any warnings like these one for any build type. I even tried to compile these files manually to make sure that all the pertaining warning flags were there. Nothing 😞

**Comment by hanhanW** [comment]
It looks like the CI uses wrong version? The bump is not applied?

Based on the commit messages, llvm version should be `1075c8ca49b467ec5e654a4322fbb36f63e8687f`, but it's `9c6a2f29660b886044a267bb4de662cd801079bc` on CI.

image

**Comment by hanhanW** [comment]
ah, they are advanced in https://github.com/google/iree/pull/9140/commits/aaac708ee0747a3e0a9e4f619778f858902ed30e

I passed all the tests in release build when pinning LLVM to `1075c8ca49b467ec5e654a4322fbb36f63e8687f` and mhlo to `b50b952f3e8e7a655c002338b4498256c75f2a07`. There are compilation failures on dbg build though.

Why do move them forward? Can we integrate it at `1075c8ca49b467ec5e654a4322fbb36f63e8687f` and handle the rest of failures in the next integration? It's always hard to fix if we pin them at an arbitrary version. It's better to pin it at one of versions that tensorflow used.

IMO, we should pin it to a version used by tensorflow and cherry-pick the needed commits. Then we don't have to handle unknown issues. but if it works at `1075c8ca49b467ec5e654a4322fbb36f63e8687f` then it's fine.

**Comment by hanhanW** [comment]
I'm going to pin LLVM and MHLO version to a working version, and see what's happening on CI. We can revert it if there are reasons to move them forward.

**Comment by dcaballe** [comment]
I initially moved forward only the llvm version to `1075c8ca49b467ec5e654a4322fbb36f63e8687f` to gather the multiple fixes for the `dyn_cast` issue. This was just a few commits ahead and I thought it would be easier than cherry-picking the commits independently. The build worked there for me after providing some fixes but it didn't working on CI. After trying a few things to solve the CI issues, since the llvm version was ahead of where it should be with respect to mhlo and tf, I decided to move all the forks forward to a new "aligned" state, just in case that was the problem. This is `aaac` and it's not bringing any new issues. `Disable clz tests for VMVX and Vulkan` was also needed in the previous state (I hadn't compile `iree-test-deps` and hadn't see those failures yet).

As I mentioned, things are working for me locally, but not on CI. I tried `Release`, `Debug` and `RelWithDebInfo` with both clang-13 and clang-6.

**Comment by dcaballe** [comment]
I fixed the buildifier issues. Feel free to move the forks as you see fit! I have the impression that I'm missing some flags to enable the tests that are failing. Fixing the warnings should be simple but I can take over tomorrow if it's not done! Thanks for looking into it!

**Comment by hanhanW** [comment]
The error in `cmake_linux_x86-swiftshader` and `buildkite/iree-build-configurations ` looks the same. Probably they are all gcc issues.

I can reproduce iree-dialect failures locally, to repro, run `cmake --build . --target check-iree-dialects`

**Comment by hanhanW** [comment]
> I decided to move all the forks forward to a new "aligned" state, just in case that was the problem.

I see, I was not aware that it's a new aligned state. Let me take a stab at it.

**Comment by hanhanW** [comment]
I feel that the error you're seeing was from previous logs. Sometimes it shows the previous log if we push commit too frequent. To re-trigger it, we can add `kokoro:force-run` label.

The bumped version has similar issues. Let's work on that. I pushed the revert commit and let's see what's happening.

I feel that we'll have to fix transform tests (which might need help from @nicolasvasilache and @matthias-springer ) and the gcc build.

**Comment by hanhanW** [comment]
I'm able to reproduce GCC failures, though I'm not sure how to fix it at this moment. Posting the repro here:

```
export CC=gcc-9
export CXX=g++-9
mkdir build
cd build
cmake -G Ninja -DIREE_ENABLE_LLD=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DIREE_BUILD_DOCS=ON -DIREE_BUILD_PYTHON_BINDINGS=ON -DIREE_ENABLE_ASSERTIONS=ON -DIREE_HAL_DRIVER_CUDA=OFF -DIREE_TARGET_BACKEND_CUDA=OFF ..
```

Errors:

```
FAILED: compiler/src/iree/compiler/Dialect/Flow/IR/FlowDialect.md /...iree/build/compiler/src/iree/compiler/Dialect/Flow/IR/FlowDialect.md
```

note: I don't have setup for cuda, so I disable it in the build. It does not matter because the errors happens in table gen stuff.

**Comment by hanhanW** [comment]
I can't reproduce CI errors either. We may fix the issues by looking into CI log. It's suspicious that it only happens on swiftshader CI. @GMNGeoffrey @ScottTodd @pzread do we have special configurations on swiftshader CI?

```mlir
ERROR: /tmpfs/src/github/iree/llvm-external-projects/iree-dialects/BUILD:277:11: Compiling llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-6.0/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG ... (remaining 212 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1595:5: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
    parser.parseAttribute(valueAttr, TileOp::getTiledDimAttrName(),
    ^~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

The rest of failures are all related to transform dialect. @nicolasvasilache could you help fix them?

```
Failed Tests (16):                                                                                                              IREE_DIALECTS :: Dialect/iree_linalg_ext/in-parallel-to-async.mlir
  IREE_DIALECTS :: Dialect/iree_linalg_ext/in-parallel-to-sequential-for.mlir
  IREE_DIALECTS :: Dialect/iree_linalg_ext/tile-to-in-parallel.mlir
  IREE_DIALECTS :: Dialect/iree_linalg_ext/tile-to-sequential-for.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/double-tiling.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/failure.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/fuse-and-peel.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/generalize.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/interchange.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/pad.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/peel.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/scalarize.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/selective-targeting.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/tile-and-peel.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/tile-interchange.mlir
  IREE_DIALECTS :: Dialect/linalg_transform/vectorize.mlir
```

File: build_tools/cmake/iree_tablegen_doc.cmake [hunk 1]
```diff
@@ -60,11 +60,16 @@ function(iree_tablegen_doc)
...
+    if(_LEN GREATER 1)
+      list(GET _RULE_OUTS 0 _DIALECT)
+      list(REMOVE_AT _RULE_OUTS 0)
+    endif()
     list(GET _RULE_OUTS 0 _OUTPUT)
     list(REMOVE_AT _RULE_OUTS 0)
 
     # TableGen this output with the given command.
-    tablegen(${_TBLGEN} ${_OUTPUT} ${_COMMAND} ${_INCLUDE_DIRS})
+    tablegen(${_TBLGEN} ${_OUTPUT} ${_COMMAND} ${_DIALECT} ${_INCLUDE_DIRS})
```
Comment (hanhanW) [inline build_tools/cmake/iree_tablegen_doc.cmake:72]: I'm not a cmake expert. Is there a better way for doing it?

All the dialects except VMVX need to add `--dialect=xxx` flag. The _RULE_OUTS is a list of either `--gen-dialect-doc -dialect=flow FlowDialect.md` or `--gen-dialect-doc VMVXDialect.md`.

We'd like to pop the last element to `_OUTPUT` and pass the rest of elements as flags to tablegen.

**Comment by hanhanW** [comment]
I'm able to reproduce the failures. This is bazel specific issues. We can follow the doc building IREE to repro it: https://github.com/google/iree/blob/main/docs/developers/get_started/building_with_bazel_linux.md

```
export CC=clang
export CXX=clang++
python configure_bazel.py
cd integrations/tensorflow
bazel test ...
```

**Comment by dcaballe** [comment]
Thanks for looking into this! 

> I'm able to reproduce the failures. This is bazel specific issues. 

How could this be a bazel only issue? The errors are due to C++ warnings. Bazel or cmake shouldn't have an impact on that unless we don't translate the flags properly between the two build systems?

Also, is there a build target that we can use to say "compile and test everything that is enabled"? Something like `check-all` in llvm or `check-mlir` in MLIR? I so far had to use different pipelines like `check-iree-dialects` or `iree-test-deps`.

**Comment by hanhanW** [comment]
> How could this be a bazel only issue?

I don't know. :) Maybe bazel has some hidden configurations.

> Also, is there a build target that we can use to say "compile and test everything that is enabled"?

I don't know either. Actually, I'm looking for it too.

**Comment by nicolasvasilache** [comment]
@hanhanW @MaheshRavishankar @dcaballe I pushed [5007844](https://github.com/google/iree/pull/9140/commits/500784443baed1711b63c62863e1ce96a16eea3d) which only uncomments commented out code.

I assume this was introduced due to compile errors with the dyn_cast issues?

Tests seem to be green now.

**Review by hanhanW — APPROVED** [review]

**Pull request merged**